flow/probes/ebpf.go

```diff
@@ -394,50 +111,31 @@ func (p *EBPFProbe) run() {
 				}
 			}
 
-			kernFlow := C.struct_flow{}
-			var key, nextKey C.__u64
+			var flowsRead int
 			for {
-				found, err := p.module.LookupNextElement(p.fmap, unsafe.Pointer(&key), unsafe.Pointer(&nextKey), unsafe.Pointer(&kernFlow))
+				found, err := p.module.LookupNextElement(p.fmap, unsafe.Pointer(&key), unsafe.Pointer(&nextKey), unsafe.Pointer(&kernFlows[nextAvailablePtr]))
 				if !found || err != nil {
+					p.module.DeleteElement(p.fmap, unsafe.Pointer(&key))
+					key = 0
 					break
 				}
-				key = nextKey
 				// delete every entry after we read the entry value
 				p.module.DeleteElement(p.fmap, unsafe.Pointer(&key))
+				key = nextKey
 
-				lastK := int64(kernFlow.last)
+				lastK := int64(kernFlows[nextAvailablePtr].last)
 				last := start.Add(time.Duration(lastK - startKTimeNs))
-
-				ebpfFlow, ok := ebpfFlows[kernFlow.key]
-				if !ok {
-					startK := int64(kernFlow.start)
-
-					// check that the local time computed from the kernel time is not greater than Now
-					us := start.Add(time.Duration(startK - startKTimeNs))
-					if us.After(now) {
-						us = now
-					}
-
-					ebpfFlow = &EBPFFlow{
-						start: now,
-						last:  last,
-					}
-					ebpfFlows[kernFlow.key] = ebpfFlow
-
-					ebpfFlow.lastK = lastK
-					ebpfFlow.last = last
-
-					for _, fops := range p.newFlowOperation(ebpfFlow, &kernFlow, startKTimeNs, start) {
-						flowChanOperation <- fops
-					}
-				} else {
-					flowChanOperation <- p.updateFlowOperation(ebpfFlow, &kernFlow, startKTimeNs, start)
+				startK := int64(kernFlows[nextAvailablePtr].start)
+				us := start.Add(time.Duration(startK - startKTimeNs))
+				if us.After(now) {
+					us = now
 				}
-			}
-
-			for k, v := range ebpfFlows {
-				if time.Now().Sub(v.last).Seconds() > p.expire.Seconds() {
-					delete(ebpfFlows, k)
+				flow.SetEBPFFlow(&ebpfFlows[nextAvailablePtr], us, last, unsafe.Pointer(&kernFlows[nextAvailablePtr]), startKTimeNs, p.probeNodeTID)
+				flowEBPFChan <- &ebpfFlows[nextAvailablePtr]
+				nextAvailablePtr = (nextAvailablePtr + 1) % flowPoolSize
+				flowsRead++
+				if flowsRead == FLOW_TABLE_SZ {
```

Comment:
don't scan more than 500K flows everytime (avoid infinite loop -- however, to avoid starvation, we now keep the index of the current flow and start from the same point next time)